### DISCUSSION TITLE
What spatial frequencies is the steerable pyramid sensitive to?

### DISCUSSION THREAD

**Discussion by billbrod** [plenoptic-org/plenoptic#323] [General]
I've had two people recently (@kbonnen and @JiyeongHa) ask me how to determine which spatial frequencies the steerable pyramid are sensitive to, in reasonable units. Right now, there's not a super straightforward way to determine that info, but the following script is an empirical way of doing so, inspired by an example in the [pyrtools docs](https://pyrtools.readthedocs.io/en/latest/tutorials/03_steerable_pyramids.html#Frequency-domain-construction):

```python
import torch
import plenoptic as po
import matplotlib.pyplot as plt
import einops

order = 3
height = 4
imsize = 512

pyr = po.simul.SteerablePyramidFreq(height=height, image_shape=[imsize, imsize], order=order)
empty_image = torch.zeros((1, 1, imsize, imsize))
pyr_coeffs = pyr.forward(empty_image)

# insert a 1 in the center of each coefficient...
for k, v in pyr.pyr_size.items():
    mid = (v[0] // 2, v[1] // 2)
    pyr_coeffs[k][0, 0, mid[0], mid[1]] = 1

# ... and then reconstruct this dummy image to visualize the filter.
reconList = []
for k in pyr_coeffs:
    # we ignore the residual_highpass and residual_lowpass, since we're focusing on the
    # filters here
    if isinstance(k, tuple):
        filt = pyr.recon_pyr(pyr_coeffs, [k[0]], [k[1]])
        filt_fourier = torch.fft.fftshift(torch.fft.fft2(filt).abs())
        reconList.append(filt_fourier)

# visualize fourier representation of filters
po.imshow(reconList, col_wrap=order + 1, vrange="indep1", zoom=2)
plt.savefig("filters.png")
plt.close()

# arrange them into a nicer shape
fourier_filts = einops.pack(reconList, "* c h w")[0]
fourier_filts = einops.rearrange(fourier_filts, "(scales ori) c h w -> scales (ori c) h w",

                                 scales=pyr.num_scales)
# sum across orientations
fourier_filts = fourier_filts.sum(1, keepdim=True)
po.imshow(fourier_filts)
plt.savefig("all_orientations.png")
plt.close()

# compute radial average
polar_rad = po.tools.polar_radius(fourier_filts.shape[-2:]).to(torch.int)
nr = torch.bincount(polar_rad.ravel())
# convert to integers so we can use bincount. amplitudes are irrelevnat
fourier_filts = (255 * fourier_filts / fourier_filts.max()).to(torch.int)
radial_avg = []
for sc in range(height):
    tbin = torch.bincount(polar_rad.ravel(), fourier_filts[sc, 0].ravel())
    radial_avg.append(tbin/nr)
![filters](https://github.com/user-attachments/assets/b14b257e-040a-4edd-88d9-c2a5f55eaf3c)

radial_avg = torch.stack(radial_avg)

plt.plot(radial_avg.T)
plt.savefig('steerpyr_freqs.png')
```

That will create three images, `filters.png`:

![filters](https://github.com/user-attachments/assets/3ef40f00-9be8-465c-b0a2-b3a81bd90726)

This shows the energy, in frequency space, of the steerable pyramid filters. Each row shows a different scale and each column a different orientation. You can see that the orientations are just rotated copies of each other, and that the frequencies each scale is selective for decreases from coarse to fine as you go up.

`all_orientations.png`:

![all_orientations](https://github.com/user-attachments/assets/ad673c07-c36d-4a3f-950e-d965db194908)

Summing across the orientations in the `filters.png` gives you the above plot, showing that each scale is an annulus, and they're just scaled copies of each other. 

and `steerpyr_freqs.png`:

![steerpyr](https://github.com/user-attachments/assets/380dcd02-8d16-4f63-8e95-bf35535d0c62)

Taking a radial average of the info in `all_orientations.png` gives you the above plot, with frequencies (in cycles per image) on the x-axis, and arbitrary units on the y.

## Deriving selectivity analytically

The steerable pyramid implementation in plenoptic is constructed in the frequency domain, so it should be possible to read this info directly out of the object itself. I am planning on refactoring the steerable pyramid at some point, and will make this info explicitly available at that point. If someone wants to get started before then, here are some pointers:

- During initialization, a bunch of frequency-domain masks are created and cached, which are used during filtering. 
- During the forward pass, we take the Fourier transform of the input image, and progressively low-pass filter/downsample for each scale.
- The actual filtering during the forward pass happens [on this line](https://github.com/plenoptic-org/plenoptic/blob/f793447e1282da6d4c5c13e09261c9df7cad479a/src/plenoptic/simulate/canonical_computations/steerable_pyramid_freq.py#L377), where `lodft` is the Fourier transformation of the image that is being progressively low-passed filtered and down-sampled on each scale, `anglemask` is the mask component that handles the different orientations (and so can be ignored for our case here), `himask` ignores all the highpass residuals, and `complex_const` can be ignored here as well.
- For our purposes, the frequency (non-angle/orientation) selectivity comes from `himask` and `lomask`, which are created during initialization and cached in the `_lomasks_scale_{i}` and `_himasks_scale_{i}` attributes (one per scale). They are off by one, so that `pyr._himasks_scale_0` should be multiplied by `pyr.lo0mask`, `pyr_himasks_scale_1` should be multiplied by `pyr._lomasks_scale_0` and so on. Unlike the `fourier_filts` variable created in the script above, these masks shrink in size with each scale, and so to get the actual frequencies, they need to be made back into images of the original size, by padding with zeros. When done correctly, the output should look a lot like the `fourier_filts` variable that has been summed across orientations.
- Those two masks are created by using some raised cosines, which all have names like `Ycosn`, `Xcosn`, etc. To understand these will require digging back into the papers describing the steerable pyramid filter design. I believe the relevant paper is [this one](https://www.cns.nyu.edu/pub/eero/karasaridis96.pdf), but I'd start with [Eero Simoncelli's notes](https://www.cns.nyu.edu/~eero/STEERPYR/) and [David Heeger's notes](https://www.cns.nyu.edu/~david/handouts/steerable.pdf). My intuition is, from looking into the arguments passed to create the raised cosine variables, we should be able to derive the exact frequencies. The advantage of this is that the above method (and similar ones) will be sensitive to sampling, whereas the math will not.
